tests/generators/test_go_generator.py

```diff
@@ -0,0 +1,1059 @@
+}""",
+        """func OpenAArchive(resource flatdata.ResourceProvider) (*AArchive, error) {
+    v := &AArchive{}
+    // Initialize resources
+	rawDataResourceIsOpen := true
+	rawDataResourceHandle, schema, err := resource.GetHandle("raw_data_resource")
+	if err != nil {
+        log.Println(err)
+	    if err.Error() == flatdata.ErrorCantAccessResource {
+	        rawDataResourceIsOpen = false
+	    } else {
+		    return v, err
+		}
+	}""",
+        """// Add resources to archive
+    v.RawDataResourceRawData = &ARawDataResourceRawData{
+        internal: rawDataResourceHandle,
+        IsOptional: false,
+        IsOpen: rawDataResourceIsOpen,
+    }
+	return v, nil
+}"""
+    ]
+
+    generate_and_assert_in("""
+    namespace n{
+    archive A {
+        raw_data_resource : raw_data;
+    }
+    }
+    """, GoGenerator, *expected_lines)
+
+
+def test_optional_resource_is_declared_correctly():
+    expected_lines = [
+        """type ARawDataResourceRawData struct {
+    internal flatdata.ResourceHandle
+    IsOptional bool
+    IsOpen bool
+}
+
+func (v *ARawDataResourceRawData) GetValue() []byte {
+	data := make([]byte, v.GetSize())
+	_, err := v.internal.ReadAt(data, 8)
+	if err != nil {
+		return make([]byte, 0)
+	}
+	return data
+}
+
+func (v *ARawDataResourceRawData) GetSize() int {
+	size := make([]byte, 8)
+	_, err := v.internal.ReadAt(size, 0)
+	if err != nil {
+		return 0
+	}
+	return int(binary.LittleEndian.Uint64(size))
+}
+
+func (v *ARawDataResourceRawData) Close() {
+    v.internal.Close()
+    v.IsOpen = false
+}
+
+func (v *ARawDataResourceRawData) GetSizeInBytes() int {
+    return v.internal.Len()
+}""",
+        """type AArchive struct {
+    IsOptional bool
+    IsOpen bool
+    RawDataResourceRawData *ARawDataResourceRawData
+}
+
+func (v *AArchive) Close() {
+    if v.RawDataResourceRawData.IsOpen {
+        v.RawDataResourceRawData.Close()
+    }
+}
+
+func (v *AArchive) GetSizeInBytes() int {
+    var size int
+    if v.RawDataResourceRawData.IsOpen {
+        size += v.RawDataResourceRawData.GetSizeInBytes()
+    }
+    return size
+}""",
+        """func OpenAArchive(resource flatdata.ResourceProvider) (*AArchive, error) {
+    v := &AArchive{}
+    // Initialize resources
+	rawDataResourceIsOpen := true
+	rawDataResourceHandle, schema, err := resource.GetHandle("raw_data_resource")
+	if err != nil {
+        log.Println(err)
+	    if err.Error() == flatdata.ErrorCantAccessResource {
+	        rawDataResourceIsOpen = false
+	    } else {
+		    return v, err
+		}
+	}""",
+        """v.RawDataResourceRawData = &ARawDataResourceRawData{
+        internal: rawDataResourceHandle,
+        IsOptional: true,
+        IsOpen: rawDataResourceIsOpen,
+    }
+	return v, nil
+}"""
+    ]
+
+    generate_and_assert_in("""
+    namespace n{
+    archive A {
+        @optional
+        raw_data_resource : raw_data;
+    }
+    }""", GoGenerator, *expected_lines)
+
+
+def test_optional_resource_is_declared_correctly():
```

Comment:
In all tests (but here is a good example) please check only what is necessary. First tests exhaustively check the code for structures and resources. So here it's enough to only check lines responsible for "optionality".

Reply:
Done